### PR TITLE
bsunanda:Run2-hcx16 Modify the Hcal parameters for geometry definition so that DDDSim/RECConstantscan be reconsturcted from this

### PR DESCRIPTION
**Pull request by bsunanda**
This is the first step for injecting new Hcal geometry definition. The Run1 and Run2 geometries in DB needs to be recreated and one can then proceed further

### CODE REVIEW

**Comment by abdoulline** [comment]
Indeed, really high-priority move, need to be integrated asap !

**Comment by mmusich** [comment]
please test

File: Geometry/HcalCommonData/interface/HcalParametersFromDD.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+class HcalParameters;
+
+class HcalParametersFromDD {
+public:
+  HcalParametersFromDD() {}
+  virtual ~HcalParametersFromDD() {}
+
+  bool build(const DDCompactView*,  HcalParameters& );
+private:
+  static const int nEtaMax=100;
```
Comment (ianna) [inline Geometry/HcalCommonData/interface/HcalParametersFromDD.h]: @bsunanda - it is better to use constexpr and since it is private, move it to .cc

**Comment by ianna** [comment]
@bsunanda - I've submitted a PR to your branch. Please, merge it.

**Comment by ianna** [comment]
-1

Need this PR to be merged: https://github.com/bsunanda/cmssw/pull/5

File: Geometry/HcalCommonData/interface/HcalGeomParameters.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+class DDFilteredView;
+
+class HcalGeomParameters {
+
+public:
+
+  HcalGeomParameters( const DDCompactView& cpv );
+  ~HcalGeomParameters();
+
+  std::vector<double>       getConstRHB()    const {return rHB;}
```
Comment (ianna) [inline Geometry/HcalCommonData/interface/HcalGeomParameters.h]: @bsunanda - you can fill in HcalParameters record directly without caching the vectors here. Copying them seems unnecessary.

**Comment by ianna** [comment]
@bsunanda - it looks like there is a new Reco on DD dependency with respect to 7_5_x:

```
7_5_x Geometry/HcalTowerAlgo:
    CondFormats/AlignmentRecord
    DataFormats/CaloTowers
    DataFormats/Common
    DataFormats/DetId
    DataFormats/GeometryVector
    DataFormats/HcalDetId
    FWCore/Framework
    FWCore/MessageLogger
    FWCore/ParameterSet
    FWCore/PythonParameterSet
    FWCore/Utilities
    Geometry/CaloGeometry
    Geometry/CaloTopology
    Geometry/Records
    tools/Boost
    tools/Roofitcore

7_6_x Geometry/HcalTowerAlgo:
    CondFormats/AlignmentRecord
    CondFormats/GeometryObjects
    DataFormats/CaloTowers
    DataFormats/Common
    DataFormats/DetId
    DataFormats/GeometryVector
    DataFormats/HcalDetId
    DetectorDescription/Base
    DetectorDescription/Core
    FWCore/Framework
    FWCore/MessageLogger
    FWCore/ParameterSet
    FWCore/PythonParameterSet
    FWCore/Utilities
    Geometry/CaloEventSetup
    Geometry/CaloGeometry
    Geometry/CaloTopology
    Geometry/HcalCommonData
    Geometry/Records
    tools/Boost
    tools/Clhepheader
    tools/Roofitcore
```

You can run the following command to check:

```
scram b dependencies
```

**Comment by bsunanda** [comment]
Hi

This is expected because HcalTowerAlgo no longer makes hardcoded geometry - it makes geometry from what is contained in xml or from DB. This is expected dependency.                     Sunanda
________________________________

**Comment by ianna** [comment]
@bsunanda - please, take https://github.com/bsunanda/cmssw/pull/6

* Avoid copying vectors
* Avoid caching vectors
* Parse DD compact view once

It reduces CPU and time from 4.77765 seconds to 3.47163 seconds (0.428102 seconds for DB)

IMHO, the remaining HcalGeomParameters code could be part of HcalParametersFromDD.

Note: you'll need to rebase your branch.

**Comment by bsunanda** [comment]
Shall I just merge with ianna:hcal-parameters-improvement - or start from a fresh version
________________________________

**Comment by civanch** [comment]
@bsunanda , if merge is working new PR is not needed.

**Comment by ianna** [comment]
@bsunanda - you can rebase your branch and then merge mine into it.

**Pull request closed without merging**